src/set.cr

```diff
@@ -311,7 +311,9 @@ struct Set(T)
   # ```
   def subtract(other : Enumerable)
     other.each do |value|
-      delete value
+      if value.is_a? T
+        delete value
+      end
```

Comment:
This seems to be unrelated to this PR, and also unnecessary.

Reply:
This is necessary, see why the previous builds failed.

Reply:
If `value` is not a `T`, `Set#delete(value)` won't return `T` and will fail to compile.

Reply:
Oh, so I think we should revert the type restriction, not add more code. You were right, it does break things.

Reply:
Actually, returning `object` is wrong. It should delete the object returned by `@hash.delete`.

Reply:
The suggestion does not really work. If the object is present, it will now return `nil` - Set is a hash of nils.

Reply:
Then I think this PR is wrong.

Reply:
Try this:

```crystal
s = Set{1, 2, 3}
s.delete(1.0) # => what does this return? It should be 1. If it returns 1.0 it's wrong.
```

Reply:
The issue will be exactly the same if we try to add a return type to `Hash#delete(key)`.

Reply:
No, because there we have the value.

Reply:
I think it makes sense to return `Bool`, not `T?`.

Reply:
In your example, `nil` will be returned. I am ok for `Bool`, no strong opinion on this.

Reply:
There is also another possible fix: add `return nil if !object.is_a? T`.

Reply:
In the case of `Array` and `Hash`, when you delete an index or a key you don't know what the underlying value is. Sometimes it's useful to know that.

In the case of `Set`, when you delete an object, that's the object that was already in the Set, so you don't learn anything new by returning that same object. It's more useful to know if the object was there or not, and for that you can use a `Bool`.

Reply:
One thing to consider is the object to ask to delete, passed to `#delete`, may not be exactly the same has the one stored.
They have the same hash of course. It may be not important after all, because it can be considered as a bad practice to have two different objects with the same hash.